tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,337 @@
+
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Helper functions to add support for iterative magnitude/gradient pruning as seen in the RiGL experiments."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from absl import logging
+
+from tensorflow.python.ops import summary_ops_v2
+from tensorflow.python.summary import summary as summary_ops_v1
+from tensorflow_model_optimization.python.core.sparsity.keras import pruning_utils
+from tensorflow_model_optimization.python.core.sparsity_tf2 import schedule as update_schedule
+from tensorflow_model_optimization.python.core.sparsity_tf2 import pruner
+from tensorflow_model_optimization.python.core.sparsity_tf2 import sparse_utils
+
+
+class RiGLPruner(pruner.Pruner):
+  """
+  Implementation of the RiGL dynamic sparse training algorithm.
+  """
+
+  def __init__(self,
+      update_schedule=update_schedule.ConstantSchedule(0.5, 0),
+      sparsity=0.5,
+      block_size=(1,1),
+      block_pooling_type='AVG',
+      initializer=sparse_utils.PermuteOnes,
+      stateless=False,
+      seed=0,
+      seed_offset=0,
+      noise_std=0,
+      reinit=False
+    ):
+    """The logic for magnitude-based pruning weight tensors.
+
+    Args:
+      update_schedule: A `PruningSchedule` object that controls pruning rate
+        throughout training.
+      sparsity: the sparsity at which the dynamic sparse training method uses
+      block_size: The dimensions (height, weight) for the block sparse pattern
+        in rank-2 weight tensors.
+      block_pooling_type: (optional) The function to use to pool weights in the
+        block. Must be 'AVG' or 'MAX'.
+      initializer: the initial sparsity distribution Callable for each layer of the network.
+      stateless: whether or not being run on TPU/multi-GPU workers.
```

Comment:
Always use stateless, it's exclusively better to have well-set up deterministic pseudo-random numbers than non-deterministic pseudo-random numbers.